src/macros.js

```diff
@@ -131,6 +136,59 @@ defineMacro("\\TextOrMath", function(context) {
     }
 });
 
+// Regular expressions for parsing numbers in base 8 through 16
+const numberRegex = {
+    "8": /^[0-7]$/,
+    "10": /^[0-9]$/,
+    "16": /^[0-9a-fA-F]$/,
+};
+
+// TeX \char makes a literal character (catcode 12) using the following forms:
+// (see The TeXBook, p. 43)
+//   \char123  -- decimal
+//   \char'123 -- octal
+//   \char"123 -- hex
+//   \char`x   -- character that can be written (i.e. isn't active)
+//   \char`\x  -- character that cannot be written (e.g. %)
+// These all refer to characters from the font, so we turn them into special
+// calls to a function \@char dealt with in the Parser.
+defineMacro("\\char", function(context) {
+    let token = context.popToken();
+    let base;
+    let number = '';
+    if (token.text === "'") {
+        base = 8;
+        token = context.popToken();
+    } else if (token.text === '"') {
+        base = 16;
+        token = context.popToken();
+    } else if (token.text === "`") {
+        token = context.popToken();
+        if (token.text[0] === "\\") {
+            number = token.text.charCodeAt(1);
+        } else if (token.text === "EOF") {
+            throw new ParseError("\\char` missing argument");
+        } else {
+            number = token.text.charCodeAt(0);
+        }
+    } else {
+        base = 10;
+    }
+    if (base) {
+        // Parse a number in the given base, starting with first `token`.
+        const regex = numberRegex[base];
+        if (!token.text.match(regex)) {
```

Comment:
You can check the digit using predefined map(`{0:0,1:1,2:2,3:3,4:4,5:5,6:6,7:7,8:8,9:9,A:10,a:10,B:11,b:11,C:12,c:12,D:13,d:13,E:14,e:14,F:15,f:15}` and checking `map[token.text] < base`. I'm not sure which is more readable.

Reply:
Hmm, maybe I could also just use `parseInt`.  `parseInt('9',8)` and `parseInt('g', 16)` both return `NaN`.  I'm not sure whether there are any other weird things allowed by `parseInt`, but it doesn't seem so.

Reply:
@edemaine `parseInt` is probably slower than `regex`.

Reply:
Good point. I'm also worried about slightly different implementations of `parseInt` on different browsers.  I switched to your map approach, with no more call to `parseInt`.